File: 3-1.py
#!/usr/bin/env python

readings = [r.strip() for r in open('3.input').readlines()]
word_len = len(readings[0])

ones = [0] * word_len
gamma = [0] * word_len
epsilon = [0] * word_len

for bit in range(word_len):
    for reading in readings:
        ones[bit] += int(reading[bit])

    if ones[bit] > len(readings) / 2:
        gamma[bit] = 1
    else:
        epsilon[bit] = 1

power = int(''.join([str(b) for b in gamma]), 2) * \
    int(''.join([str(b) for b in epsilon]), 2)
print(power)

File: 3-1.rb
#!/usr/bin/env ruby

readings = File.read('3.input').lines.map(&:strip)
word_len = readings[0].size

ones = [0] * word_len
gamma = [0] * word_len
epsilon = [0] * word_len

0.upto(word_len - 1) do |bit|
  readings.each do |reading|
    ones[bit] += reading[bit].to_i
  end

  if ones[bit] > readings.size / 2
    gamma[bit] = 1
  else
    epsilon[bit] = 1
  end
end

power = gamma.join.to_i(2) * epsilon.join.to_i(2)

print power, "\n"
